openpype/hosts/maya/plugins/publish/extract_import_reference.py

```diff
@@ -0,0 +1,156 @@
+
+                except KeyError:
+                    # set scene type to ma
+                    self.scene_type = "ma"
+
+        _scene_type = ("mayaAscii"
+                       if self.scene_type == "ma"
+                       else "mayaBinary")
+
+        dir_path = self.staging_dir(instance)
+        # named the file with imported reference
+        if instance.name == "Main":
+            return
+        tmp_name = instance.name + self.tmp_format
+        current_name = cmds.file(query=True, sceneName=True)
+        ref_scene_name = "{0}.{1}".format(tmp_name, self.scene_type)
+
+        reference_path = os.path.join(dir_path, ref_scene_name)
+        tmp_path = os.path.dirname(current_name) + "/" + ref_scene_name
+
+        self.log.info("Performing extraction..")
+
+        # This generates script for mayapy to take care of reference
+        # importing outside current session. It is passing current scene
+        # name and destination scene name.
+        script = ("""
+# -*- coding: utf-8 -*-
+'''Script to import references to given scene.'''
+import maya.standalone
+maya.standalone.initialize()
+# scene names filled by caller
+current_name = "{current_name}"
+ref_scene_name = "{ref_scene_name}"
+print(">>> Opening {{}} ...".format(current_name))
+cmds.file(current_name, open=True, force=True)
+print(">>> Processing references")
+all_reference = cmds.file(q=True, reference=True) or []
+for ref in all_reference:
+    if cmds.referenceQuery(ref, il=True):
+        cmds.file(ref, importReference=True)
+
+        nested_ref = cmds.file(q=True, reference=True)
+        if nested_ref:
+            for new_ref in nested_ref:
+                if new_ref not in all_reference:
+                    all_reference.append(new_ref)
+
+print(">>> Finish importing references")
+print(">>> Saving scene as {{}}".format(ref_scene_name))
+
+cmds.file(rename=ref_scene_name)
+cmds.file(save=True, force=True)
+print("*** Done")
+        """).format(current_name=current_name, ref_scene_name=tmp_path)
+        mayapy_exe = os.path.join(os.getenv("MAYA_LOCATION"), "bin", "mayapy")
+        if sys.platform == "windows":
+            mayapy_exe += ".exe"
+            mayapy_exe = os.path.normpath(mayapy_exe)
+        # can't use TemporaryNamedFile as that can't be opened in another
+        # process until handles are closed by context manager.
+        with tempfile.TemporaryDirectory() as tmp_dir_name:
+            tmp_file_name = os.path.join(tmp_dir_name, "import_ref.py")
+            tmp = open(tmp_file_name, "w+t")
+            subprocess_args = [
+                mayapy_exe,
+                tmp_file_name
+            ]
+            self.log.info("Using temp file: {}".format(tmp.name))
+            try:
+                tmp.write(script)
+                tmp.close()
+                run_subprocess(subprocess_args)
+            except Exception:
+                self.log.error("Import reference failed", exc_info=True)
+                raise
```

Comment:
Why are we opening the file with `+` mode? Why is that required?

I feel this would make for more readable code (note that this doesn't capture the script file writing in try..except like before). What do you think?
```suggestion
            tmp_script_path = os.path.join(tmp_dir_name, "import_ref.py")
            self.log.info("Using script file: {}".format(tmp_script_path))
            with open(tmp_script_path, "wt") as tmp:
                tmp.write(script)
                
            try:
                run_subprocess([mayapy_exe, tmp_script_path])
            except Exception:
                self.log.error("Import reference failed", exc_info=True)
                raise
```